tutorials/langchain_google_palm_tracing_tutorial.ipynb

```diff
@@ -0,0 +1,230 @@
+    "        |\n",
+    "        <a href=\"https://join.slack.com/t/arize-ai/shared_invite/zt-1px8dcmlf-fmThhDFD_V_48oU7ALan4Q\">Community</a>\n",
+    "    </p>\n",
+    "</center>\n",
+    "<h1 align=\"center\">Tracing a LangChain + Google PaLM Application</h1>\n",
+    "\n",
+    "LLM orchestration frameworks such as LangChain provide abstractions that enable users to build powerful applications in a few lines of code. However, the same abstractions make it difficult to understand what is going on under the hood and to pinpoint the cause of issues.\n",
+    "\n",
+    "Phoenix makes your LLM applications *observable* by providing tracing.\n",
+    "\n",
+    "In this tutorial, you will:\n",
+    "- Build a simple LLM application using LangChain and Google PaLM,\n",
+    "- Record trace data in OpenInference format,\n",
+    "- Inspect the traces and spans of your application to identify uncaught exceptions and sources of latency and cost.\n",
+    "\n",
+    "ℹ️️ This notebook requires access to the [Google PaLM API](https://developers.generativeai.google/) and a Google API key.\n",
+    "\n",
+    "❗ Phoenix tracing is in alpha. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Install Dependencies and Import Libraries"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Install Phoenix and LangChain."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -qq \"arize-phoenix[experimental]==0.0.33rc2\" google-generativeai langchain"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Import libraries."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import os\n",
+    "from getpass import getpass\n",
+    "from urllib.request import urlopen\n",
+    "\n",
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "import phoenix as px\n",
+    "from langchain.chains import RetrievalQA\n",
+    "from langchain.chat_models import ChatGooglePalm\n",
+    "from langchain.embeddings import GooglePalmEmbeddings\n",
+    "from langchain.retrievers import KNNRetriever\n",
+    "from phoenix.experimental.callbacks.langchain_tracer import OpenInferenceTracer\n",
+    "from phoenix.trace.trace_dataset import TraceDataset\n",
+    "from tqdm import tqdm"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Configure Your Google API Key"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set your Google API key."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "google_api_key = os.environ.get(\"GOOGLE_API_KEY\", getpass(\"Enter your Google API Key: \"))\n",
```

Comment:
this is actually a bug because the second argument is _always_ evaluated.
```suggestion
    "if not (google_api_key := os.getenv(\"GOOGLE_API_KEY\")):\n",
    "    google_api_key = getpass(\"Enter your Google API Key: \")\n",
```

Reply:
Good catch.

Reply:
addressed in 894d7fe